### PR TITLE
Improve performance in the lots-of-simulations case

### PR DESCRIPTION
**Pull request by jeromekelleher**
Second pass at this - in retropect, hacking in dodgy hash definitions is a terrible idea. It's not worth the convenience of the lru_cache decorator.

@grahamgower, can you take a quick look please? Performance is the same as the other version.

### CODE REVIEW

File: msprime/demography.py [hunk 1]
```diff
@@ -58,6 +59,76 @@ class IncompletePopulationMetadataWarning(UserWarning):
...
+
+def _build_population_table(populations):
+    """
+    Return a tskit PopulationTable instance encoding the metadata for the
+    specified populations. Because encoding metadata is quite expensive
+    we maintain an LRU cache.
+    """
+    key = json.dumps([pop.asdict() for pop in populations], sort_keys=True)
+    if key not in _population_table_cache:
+        # if True:
```
Comment (grahamgower) [inline msprime/demography.py:91]: Leftover from debugging?

**Review by grahamgower — APPROVED** [review]
LGTM!

File: msprime/demography.py (hunk 1 above)
Comment (jeromekelleher) [reply msprime/demography.py:91]: :facepalm:

File: msprime/demography.py [hunk 2]
```diff
@@ -58,6 +59,75 @@ class IncompletePopulationMetadataWarning(UserWarning):
...
+_population_table_cache = LruCache(16)
+
+
+def _build_population_table(populations):
+    """
+    Return a tskit PopulationTable instance encoding the metadata for the
+    specified populations. Because encoding metadata is quite expensive
+    we maintain an LRU cache.
+    """
+    key = json.dumps([pop.asdict() for pop in populations], sort_keys=True)
```
Comment (grahamgower) [inline msprime/demography.py:89]: Hang on, doesn't Population.asdict() include demographic information, not just name/description/metadata as used here?

**Comment by jeromekelleher** [comment]
Great, thanks @grahamgower! I'll merge and push out 1.0.3 ASAP

File: msprime/demography.py (hunk 2 above)
Comment (jeromekelleher) [reply msprime/demography.py:89]: Yes, you're right - I guess if you want to make the minimal key you'd just build the actual population metadata items and cache on them.

Sigh - I'll update.

Comment (jeromekelleher) [reply msprime/demography.py]: OK, updated - how does it look now?

Comment (grahamgower) [reply msprime/demography.py]: Yes, that looks right now. Thanks!

**Comment by jeromekelleher** [comment]
OK, merging. Thanks for your help @grahamgower!

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,9 @@
 - Support for running full ARG simulations with gene conversion
   ({pr}`1801`, {issue}`1773`, {user}`JereKoskela`).
 
+- Improved performance when running many small simulations
+  ({pr}`1909`, {user}`jeromekelleher`.
```
Comment (benjeffery) [inline CHANGELOG.md:11]: Missing a bracket

**Review by benjeffery — COMMENTED** [review]
LGTM, I think tskit should be doing some of the caching if possible though.

**Pull request merged**